File: typescript/standard/01-agent-lifecycle.ts
/**
 * Standard API Example 01: Agent Lifecycle
 *
 * Demonstrates the full Agent lifecycle: start, pause, resume, stop.
 * The Agent class provides more control than Basic API's provide().
 *
 * This example demonstrates:
 * - Agent construction and configuration
 * - Lifecycle methods: start(), pause(), resume(), stop()
 * - Status tracking and events
 *
 * Run: npx ts-node standard/01-agent-lifecycle.ts
 */

import { Agent, request, serviceDirectory } from '@agirails/sdk';
import { sleep } from '../src/utils/helpers';

async function main() {
  console.log('AGIRAILS Standard API - Agent Lifecycle\n');

  // 1. Create an Agent with configuration
  const agent = new Agent({
    name: 'LifecycleDemo',
    description: 'Demonstrates agent lifecycle management',
    network: 'mock',
    behavior: {
      autoAccept: true,
      concurrency: 5, // Max 5 concurrent jobs
    },
    logging: {
      level: 'info',
    },
  });

  console.log(`Agent created: ${agent.name}`);
  console.log(`Status: ${agent.status}`); // 'idle'

  // 2. Register a service
  agent.provide('ping', async (job) => {
    return { pong: true, timestamp: Date.now() };
  });

  console.log(`Services: ${agent.serviceNames.join(', ')}`);

  // 3. Start the agent
  console.log('\nStarting agent...');
  await agent.start();

  // Register with service directory (needed for request() to find this agent)
  serviceDirectory.register('ping', agent.address);

  console.log(`Status: ${agent.status}`); // 'running'
  console.log(`Address: ${agent.address}`);

  // 4. Make a request while running
  console.log('\nMaking request while agent is running...');
  const { result: result1 } = await request('ping', { input: {}, budget: 0.10 });
  console.log('Response:', result1);

  // 5. Pause the agent
  console.log('\nPausing agent...');
  agent.pause();
  console.log(`Status: ${agent.status}`); // 'paused'

  // Agent won't accept new jobs while paused
  // (existing jobs will complete)

  // 6. Resume the agent
  console.log('\nResuming agent...');
  agent.resume();
  console.log(`Status: ${agent.status}`); // 'running'

  // 7. Make another request
  console.log('\nMaking request after resume...');
  const { result: result2 } = await request('ping', { input: {}, budget: 0.10 });
  console.log('Response:', result2);

  // 8. Show final stats
  console.log('\nAgent Stats:');
  console.log(`  Jobs completed: ${agent.stats.jobsCompleted}`);
  console.log(`  Total earned: $${agent.stats.totalEarned.toFixed(2)}`);
  console.log(`  Success rate: ${(agent.stats.successRate * 100).toFixed(0)}%`);

  // 9. Stop the agent
  console.log('\nStopping agent...');
  serviceDirectory.unregister('ping', agent.address);
  await agent.stop();
  console.log(`Status: ${agent.status}`); // 'stopped'

  console.log('\nLifecycle demo complete!');
}

main().catch(console.error);

File: python/standard/01_agent_lifecycle.py
#!/usr/bin/env python3
"""
Standard API Example 01: Agent Lifecycle

Demonstrates the full Agent lifecycle: start, pause, resume, stop.
The Agent class provides more control than Basic API's provide().

This example demonstrates:
- Agent construction and configuration
- Lifecycle methods: start(), pause(), resume(), stop()
- Status tracking and events

Run: python standard/01_agent_lifecycle.py
"""

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from src.utils.helpers import clear_mock_state


async def main() -> None:
    print("AGIRAILS Standard API - Agent Lifecycle\n")

    clear_mock_state()

    from agirails.level1 import Agent, AgentConfig, AgentBehavior, Job, JobContext

    # 1. Create an Agent with configuration
    agent = Agent(
        AgentConfig(
            name="LifecycleDemo",
            description="Demonstrates agent lifecycle management",
            network="mock",
            behavior=AgentBehavior(
                auto_accept=True,
                concurrency=5,  # Max 5 concurrent jobs
            ),
        )
    )

    print(f"Agent created: {agent.name}")
    print(f"Status: {agent.status.value}")  # 'idle'

    # 2. Register a service using decorator
    @agent.provide("ping")
    async def handle_ping(job: Job, ctx: JobContext) -> dict:
        ctx.log.info("Processing ping request")
        return {"pong": True, "message": "Hello from agent!"}

    print(f"Services: {', '.join(agent.service_names)}")

    # 3. Start the agent
    print("\nStarting agent...")
    await agent.start()

    print(f"Status: {agent.status.value}")  # 'running'
    print(f"Address: {agent.address}")

    # In mock mode, the agent handles jobs internally
    # In production, jobs come from blockchain events

    # 4. Demonstrate lifecycle operations
    await asyncio.sleep(1)

    # 5. Pause the agent
    print("\nPausing agent...")
    agent.pause()
    print(f"Status: {agent.status.value}")  # 'paused'

    # Agent won't accept new jobs while paused
    await asyncio.sleep(1)

    # 6. Resume the agent
    print("\nResuming agent...")
    agent.resume()
    print(f"Status: {agent.status.value}")  # 'running'

    await asyncio.sleep(1)

    # 7. Show stats
    print("\nAgent Stats:")
    print(f"  Jobs received: {agent.stats.jobs_received}")
    print(f"  Jobs completed: {agent.stats.jobs_completed}")
    print(f"  Total earned: ${agent.stats.total_earned:.2f}")
    print(f"  Success rate: {agent.stats.success_rate:.0f}%")

    # 8. Balance info
    print("\nAgent Balance:")
    print(f"  ETH: {agent.balance.eth}")
    print(f"  USDC: {agent.balance.usdc}")

    # 9. Stop the agent
    print("\nStopping agent...")
    await agent.stop()
    print(f"Status: {agent.status.value}")  # 'stopped'

    print("\nLifecycle demo complete!")


if __name__ == "__main__":
    asyncio.run(main())
